README.rst
UpSetPlot documentation
============================

|version| |licence| |py-versions|

|issues| |build| |docs| |coverage|

This is another Python implementation of UpSet plots by Lex et al. [Lex2014]_.
UpSet plots are used to visualise set overlaps; like Venn diagrams but
more readable. Documentation is at https://upsetplot.readthedocs.io.

This ``upsetplot`` library tries to provide a simple interface backed by an
extensible, object-oriented design.

There are many ways to represent the categorisation of data, as covered in
our `Data Format Guide <https://upsetplot.readthedocs.io/en/stable/formats.html>`_.

Our internal input format uses a `pandas.Series` containing counts
corresponding to subset sizes, where each subset is an intersection of named
categories.  The index of the Series indicates which rows pertain to which
categories, by having multiple boolean indices, like ``example`` in the
following::

    >>> from upsetplot import generate_counts
    >>> example = generate_counts()
    >>> example
    cat0   cat1   cat2
    False  False  False      56
                  True      283
           True   False    1279
                  True     5882
    True   False  False      24
                  True       90
           True   False     429
                  True     1957
    Name: value, dtype: int64

Then::

    >>> from upsetplot import plot
    >>> plot(example)  # doctest: +SKIP
    >>> from matplotlib import pyplot
    >>> pyplot.show()  # doctest: +SKIP

makes:

.. image:: http://upsetplot.readthedocs.io/en/latest/_images/sphx_glr_plot_generated_001.png
   :target: ../auto_examples/plot_generated.html

And you can save the image in various formats::

    >>> pyplot.savefig("/path/to/myplot.pdf")  # doctest: +SKIP
    >>> pyplot.savefig("/path/to/myplot.png")  # doctest: +SKIP

This plot shows the cardinality of every category combination seen in our data.
The leftmost column counts items absent from any category. The next three
columns count items only in ``cat1``, ``cat2`` and ``cat3`` respectively, with
following columns showing cardinalities for items in each combination of
exactly two named sets. The rightmost column counts items in all three sets.

Rotation
........

We call the above plot style "horizontal" because the category intersections
are presented from left to right.  `Vertical plots
<http://upsetplot.readthedocs.io/en/latest/auto_examples/plot_vertical.html>`__
are also supported!

.. image:: http://upsetplot.readthedocs.io/en/latest/_images/sphx_glr_plot_vertical_001.png
   :target: http://upsetplot.readthedocs.io/en/latest/auto_examples/plot_vertical.html

Distributions
.............

Providing a DataFrame rather than a Series as input allows us to expressively
`plot the distribution of variables
<http://upsetplot.readthedocs.io/en/latest/auto_examples/plot_diabetes.html>`__
in each subset.

.. image:: http://upsetplot.readthedocs.io/en/latest/_images/sphx_glr_plot_diabetes_001.png
   :target: http://upsetplot.readthedocs.io/en/latest/auto_examples/plot_diabetes.html

Loading datasets
................

While the dataset above is randomly generated, you can prepare your own dataset
for input to upsetplot.  A helpful tool is `from_memberships`, which allows
us to reconstruct the example above by indicating each data point's category
membership::

    >>> from upsetplot import from_memberships
    >>> example = from_memberships(
    ...     [[],
    ...      ['cat2'],
    ...      ['cat1'],
    ...      ['cat1', 'cat2'],
    ...      ['cat0'],
    ...      ['cat0', 'cat2'],
    ...      ['cat0', 'cat1'],
    ...      ['cat0', 'cat1', 'cat2'],
    ...      ],
    ...      data=[56, 283, 1279, 5882, 24, 90, 429, 1957]
    ... )
    >>> example
    cat0   cat1   cat2
    False  False  False      56
                  True      283
           True   False    1279
                  True     5882
    True   False  False      24
                  True       90
           True   False     429
                  True     1957
    dtype: int64

See also `from_contents`, another way to describe categorised data, and
`from_indicators` which allows each category to be indicated by a column in
the data frame (or a function of the column's data such as whether it is a
missing value).

Installation
------------

To install the library, you can use `pip`::

    $ pip install upsetplot

Installation requires:

* pandas
* matplotlib >= 2.0
* seaborn to use `UpSet.add_catplot`

It should then be possible to::

    >>> import upsetplot

in Python.

Why an alternative to py-upset?
-------------------------------

Probably for petty reasons. It appeared `py-upset
<https://github.com/ImSoErgodic/py-upset>`_ was not being maintained.  Its
input format was undocumented, inefficient and, IMO, inappropriate.  It did not
facilitate showing plots of each subset's distribution as in Lex et al's work
introducing UpSet plots. Nor did it include the horizontal bar plots
illustrated there. It did not support Python 2. I decided it would be easier to
construct a cleaner version than to fix it.

References
----------

.. [Lex2014] Alexander Lex, Nils Gehlenborg, Hendrik Strobelt, Romain Vuillemot, Hanspeter Pfister,
   *UpSet: Visualization of Intersecting Sets*,
   IEEE Transactions on Visualization and Computer Graphics (InfoVis '14), vol. 20, no. 12, pp. 1983–1992, 2014.
   doi: `doi.org/10.1109/TVCG.2014.2346248 <https://doi.org/10.1109/TVCG.2014.2346248>`_


.. |py-versions| image:: https://img.shields.io/pypi/pyversions/upsetplot.svg
    :alt: Python versions supported

.. |version| image:: https://badge.fury.io/py/UpSetPlot.svg
    :alt: Latest version on PyPi
    :target: https://badge.fury.io/py/UpSetPlot

.. |build| image:: https://github.com/jnothman/upsetplot/actions/workflows/test.yml/badge.svg
    :alt: Github Workflows CI build status
    :scale: 100%
    :target: https://github.com/jnothman/UpSetPlot/actions/workflows/test.yml

.. |issues| image:: https://img.shields.io/github/issues/jnothman/UpSetPlot.svg
    :alt: Issue tracker
    :target: https://github.com/jnothman/UpSetPlot

.. |coverage| image:: https://coveralls.io/repos/github/jnothman/UpSetPlot/badge.svg
    :alt: Test coverage
    :target: https://coveralls.io/github/jnothman/UpSetPlot

.. |docs| image:: https://readthedocs.org/projects/upsetplot/badge/?version=latest
     :alt: Documentation Status
     :scale: 100%
     :target: https://upsetplot.readthedocs.io/en/latest/?badge=latest

.. |licence| image:: https://img.shields.io/badge/Licence-BSD-blue.svg
     :target: https://opensource.org/licenses/BSD-3-Clause


CHANGELOG.rst
What's new in version 1.0
-------------------------

In development

What's new in version 0.9
-------------------------

- Fixes a bug where ``show_percentages`` used the incorrect denominator if
  filtering (e.g. ``min_subset_size``) was applied. This bug was a regression
  introduced in version 0.7. (:issue:`248`)
- Align ylabels of subplots added using ``add_catplot``. (:issue:`266`)
- Add a ``style_categories`` method to customize category plot styles, including
  shading of rows in the intersection matrix, and bars in the totals plot.
  (:issue:`261` with thanks to :user:`Marcel Albus <maralbus>`).
- Ability to disable totals plot with `totals_plot_elements=0`. (:issue:`246`)
- Ability to set totals y axis label (:issue:`243`)
- Added ``max_subset_rank`` to get only n most populous subsets. (:issue:`253`)
- Added support for ``min_subset_size`` and ``max_subset_size`` specified as
  percentage. (:issue:`264`)

What's new in version 0.8
-------------------------

- Allowed ``show_percentages`` to be provided with a custom formatting string,
  for example to show more decimal places. (:issue:`194`)
- Added `include_empty_subsets` to `UpSet` and `query` to allow the display of
  all possible subsets. (:issue:`185`)
- `sort_by` and `sort_categories_by` now accept '-' prefix to their values
  to sort in reverse. 'input' and '-input' are also supported. (:issue:`180`)
- Added `subsets` attribute to QueryResult. (:issue:`198`)
- Fixed a bug where more than 64 categories could result in an error. (:issue:`193`)

Patch release 0.8.2 handles deprecations in dependencies.

What's new in version 0.7
-------------------------

- Added `query` function to support analysing set-based data.
- Fixed support for matplotlib >3.5.2 (:issue:`191`. Thanks :user:`GuyTeichman`)

What's new in version 0.6
-------------------------

- Added `add_stacked_bars`, similar to `add_catplot` but to add stacked bar
  charts to show discrete variable distributions within each subset.
  (:issue:`137`)
- Improved ability to control colors, and added a new example of same.
  Parameters ``other_dots_color`` and ``shading_color`` were added.
  ``facecolor`` will now default to white if
  ``matplotlib.rcParams['axes.facecolor']`` is dark. (:issue:`138`)
- Added `style_subsets` to colour intersection size bars and matrix
  dots in the plot according to a specified query. (:issue:`152`)
- Added `from_indicators` to allow yet another data input format. This
  allows category membership to be easily derived from a DataFrame, such as
  when plotting missing values in the columns of a DataFrame. (:issue:`143`)

What's new in version 0.5
-------------------------

- Support using input intersection order with ``sort_by=None`` (:issue:`133`
  with thanks to :user:`Brandon B <outlace>`).
- Add parameters for filtering by subset size (with thanks to
  :user:`Sichong Peng <SichongP>`) and degree. (:issue:`134`)
- Fixed an issue where tick labels were not given enough space and overlapped
  category totals. (:issue:`132`)
- Fixed an issue where our implementation of ``sort_by='degree'`` apparently
  gave incorrect results for some inputs and versions of Pandas. (:issue:`134`)

What's new in version 0.4.4
---------------------------

- Fixed a regresion which caused the first column to be hidden
  (:issue:`125`)

What's new in version 0.4.3
---------------------------

- Fixed issue with the order of catplots being reversed for vertical plots
  (:issue:`122` with thanks to :user:`Enrique Fernandez-Blanco <ennanco>`)
- Fixed issue with the x limits of vertical plots (:issue:`121`).

What's new in version 0.4.2
---------------------------

- Fixed large x-axis plot margins with high number of unique intersections
  (:issue:`106` with thanks to :user:`Yidi Huang <huangy6>`)

What's new in version 0.4.1
---------------------------

- Fixed the calculation of percentage which was broken in 0.4.0. (:issue:`101`)

What's new in version 0.4
-------------------------

- Added option to display both the absolute frequency and the percentage of
  the total for each intersection and category. (:issue:`89` with thanks to
  :user:`Carlos Melus <maziello>` and :user:`Aaron Rosenfeld <arosenfeld>`)
- Improved efficiency where there are many categories, but valid combinations
  are sparse, if `sort_by='degree'`. (:issue:`82`)
- Permit truthy (not necessarily bool) values in index.
  (:issue:`74` with thanks to :user:`ZaxR`)
- `intersection_plot_elements` can now be set to 0 to hide the intersection
  size plot when `add_catplot` is used. (:issue:`80`)

What's new in version 0.3
-------------------------

- Added `from_contents` to provide an alternative, intuitive way of specifying
  category membership of elements.
- To improve code legibility and intuitiveness, `sum_over=False` was deprecated
  and a `subset_size` parameter was added.  It will have better default
  handling of DataFrames after a short deprecation period.
- `generate_data` has been replaced with `generate_counts` and
  `generate_samples`.
- Fixed the display of the "intersection size" label on plots, which had been
  missing.
- Trying to improve nomenclature, upsetplot now avoids "set" to refer to the
  top-level sets, which are now to be known as "categories". This matches the
  intuition that categories are named, logical groupings, as opposed to
  "subsets". To this end:

  - `generate_counts` (formerly `generate_data`) now names its categories
    "cat1", "cat2" etc. rather than "set1", "set2", etc.
  - the `sort_sets_by` parameter has been renamed to `sort_categories_by` and
    will be removed in version 0.4.

What's new in version 0.2.1
---------------------------

- Return a Series (not a DataFrame) from `from_memberships` if data is
  1-dimensional.

What's new in version 0.2
-------------------------

- Added `from_memberships` to allow a more convenient data input format.
- `plot` and `UpSet` now accept a `pandas.DataFrame` as input, if the

upsetplot/reformat.py
import typing

import numpy as np
import pandas as pd


def _aggregate_data(df, subset_size, sum_over):
    """
    Returns
    -------
    df : DataFrame
        full data frame
    aggregated : Series
        aggregates
    """
    _SUBSET_SIZE_VALUES = ["auto", "count", "sum"]
    if subset_size not in _SUBSET_SIZE_VALUES:
        raise ValueError(
            f"subset_size should be one of {_SUBSET_SIZE_VALUES}."
            f" Got {repr(subset_size)}"
        )
    if df.ndim == 1:
        # Series
        input_name = df.name
        df = pd.DataFrame({"_value": df})

        if subset_size == "auto" and not df.index.is_unique:
            raise ValueError(
                'subset_size="auto" cannot be used for a '
                "Series with non-unique groups."
            )
        if sum_over is not None:
            raise ValueError("sum_over is not applicable when the input is a " "Series")
        sum_over = False if subset_size == "count" else "_value"
    else:
        # DataFrame
        if sum_over is False:
            raise ValueError("Unsupported value for sum_over: False")
        elif subset_size == "auto" and sum_over is None:
            sum_over = False
        elif subset_size == "count":
            if sum_over is not None:
                raise ValueError(
                    "sum_over cannot be set if subset_size=%r" % subset_size
                )
            sum_over = False
        elif subset_size == "sum" and sum_over is None:
            raise ValueError(
                "sum_over should be a field name if "
                'subset_size="sum" and a DataFrame is '
                "provided."
            )

    gb = df.groupby(level=list(range(df.index.nlevels)), sort=False)
    if sum_over is False:
        aggregated = gb.size()
        aggregated.name = "size"
    elif hasattr(sum_over, "lower"):
        aggregated = gb[sum_over].sum()
    else:
        raise ValueError("Unsupported value for sum_over: %r" % sum_over)

    if aggregated.name == "_value":
        aggregated.name = input_name

    return df, aggregated


def _check_index(df):
    # check all indices are boolean
    if not all({True, False} >= set(level) for level in df.index.levels):
        raise ValueError(
            "The DataFrame has values in its index that are not " "boolean"
        )
    df = df.copy(deep=False)
    # XXX: this may break if input is not MultiIndex
    kw = {
        "levels": [x.astype(bool) for x in df.index.levels],
        "names": df.index.names,
    }
    if hasattr(df.index, "codes"):
        # compat for pandas <= 0.20
        kw["codes"] = df.index.codes
    else:
        kw["labels"] = df.index.labels
    df.index = pd.MultiIndex(**kw)
    return df


def _scalar_to_list(val):
    if not isinstance(val, (typing.Sequence, set)) or isinstance(val, str):
        val = [val]
    return val


def _check_percent(value, agg):
    if not isinstance(value, str):
        return value
    try:
        if value.endswith("%") and 0 <= float(value[:-1]) <= 100:
            return float(value[:-1]) / 100 * agg.sum()
    except ValueError:
        pass
    raise ValueError(
        f"String value must be formatted as percentage between 0 and 100. Got {value}"
    )


def _get_subset_mask(
    agg,
    min_subset_size,
    max_subset_size,
    max_subset_rank,
    min_degree,
    max_degree,
    present,
    absent,
):
    """Get a mask over subsets based on size, degree or category presence"""
    min_subset_size = _check_percent(min_subset_size, agg)
    max_subset_size = _check_percent(max_subset_size, agg)
    subset_mask = True
    if min_subset_size is not None:
        subset_mask = np.logical_and(subset_mask, agg >= min_subset_size)
    if max_subset_size is not None:
        subset_mask = np.logical_and(subset_mask, agg <= max_subset_size)
    if max_subset_rank is not None:
        subset_mask = np.logical_and(
            subset_mask, agg.rank(method="min", ascending=False) <= max_subset_rank
        )
    if (min_degree is not None and min_degree >= 0) or max_degree is not None:
        degree = agg.index.to_frame().sum(axis=1)
        if min_degree is not None:
            subset_mask = np.logical_and(subset_mask, degree >= min_degree)
        if max_degree is not None:
            subset_mask = np.logical_and(subset_mask, degree <= max_degree)
    if present is not None:
        for col in _scalar_to_list(present):
            subset_mask = np.logical_and(
                subset_mask, agg.index.get_level_values(col).values
            )
    if absent is not None:
        for col in _scalar_to_list(absent):
            exclude_mask = np.logical_not(agg.index.get_level_values(col).values)
            subset_mask = np.logical_and(subset_mask, exclude_mask)
    return subset_mask


def _filter_subsets(
    df,
    agg,
    min_subset_size,
    max_subset_size,
    max_subset_rank,
    min_degree,
    max_degree,
    present,
    absent,
):
    subset_mask = _get_subset_mask(
        agg,
        min_subset_size=min_subset_size,
        max_subset_size=max_subset_size,
        max_subset_rank=max_subset_rank,
        min_degree=min_degree,
        max_degree=max_degree,
        present=present,
        absent=absent,
    )

    if subset_mask is True:
        return df, agg

    agg = agg[subset_mask]
    df = df[df.index.isin(agg.index)]
    return df, agg


class QueryResult:
    """Container for reformatted data and aggregates

    Attributes
    ----------
    data : DataFrame
        Selected samples. The index is a MultiIndex with one boolean level for
        each category.
    subsets : dict[frozenset, DataFrame]
        Dataframes for each intersection of categories.
    subset_sizes : Series
        Total size of each selected subset as a series. The index is as
        for `data`.
    category_totals : Series
        Total size of each category, regardless of selection.
    total : number
        Total number of samples, or sum of sum_over value.
    """

    def __init__(self, data, subset_sizes, category_totals, total):
        self.data = data
        self.subset_sizes = subset_sizes
        self.category_totals = category_totals
        self.total = total

    def __repr__(self):
        return (
            "QueryResult(data={data}, subset_sizes={subset_sizes}, "
            "category_totals={category_totals}, total={total}".format(**vars(self))
        )

    @property
    def subsets(self):
        categories = np.asarray(self.data.index.names)
        return {
            frozenset(categories.take(mask)): subset_data
            for mask, subset_data in self.data.groupby(
                level=list(range(len(categories))), sort=False
            )
        }


def query(
    data,
    present=None,
    absent=None,
    min_subset_size=None,
    max_subset_size=None,
    max_subset_rank=None,
    min_degree=None,
    max_degree=None,
    sort_by="degree",
    sort_categories_by="cardinality",
    subset_size="auto",
    sum_over=None,
    include_empty_subsets=False,
):
    """Transform and filter a categorised dataset

    Retrieve the set of items and totals corresponding to subsets of interest.

    Parameters
    ----------
    data : pandas.Series or pandas.DataFrame
        Elements associated with categories (a DataFrame), or the size of each
        subset of categories (a Series).
        Should have MultiIndex where each level is binary,
        corresponding to category membership.
        If a DataFrame, `sum_over` must be a string or False.
    present : str or list of str, optional
        Category or categories that must be present in subsets for styling.
    absent : str or list of str, optional
        Category or categories that must not be present in subsets for
        styling.
    min_subset_size : int or "number%", optional
        Minimum size of a subset to be reported. All subsets with
        a size smaller than this threshold will be omitted from
        category_totals and data.  This may be specified as a percentage
        using a string, like "50%".
        Size may be a sum of values, see `subset_size`.

        .. versionchanged:: 0.9
            Support percentages
    max_subset_size : int or "number%", optional
        Maximum size of a subset to be reported.

        .. versionchanged:: 0.9
            Support percentages
    max_subset_rank : int, optional
        Limit to the top N ranked subsets in descending order of size.
        All tied subsets are included.

        .. versionadded:: 0.9
    min_degree : int, optional
        Minimum degree of a subset to be reported.
    max_degree : int, optional
        Maximum degree of a subset to be reported.
    sort_by : {'cardinality', 'degree', '-cardinality', '-degree',
               'input', '-input'}
        If 'cardinality', subset are listed from largest to smallest.
        If 'degree', they are listed in order of the number of categories
        intersected. If 'input', the order they appear in the data input is
        used.
        Prefix with '-' to reverse the ordering.

        Note this affects ``subset_sizes`` but not ``data``.
    sort_categories_by : {'cardinality', '-cardinality', 'input', '-input'}
        Whether to sort the categories by total cardinality, or leave them
        in the input data's provided order (order of index levels).
        Prefix with '-' to reverse the ordering.
    subset_size : {'auto', 'count', 'sum'}
        Configures how to calculate the size of a subset. Choices are:

        'auto' (default)
            If `data` is a DataFrame, count the number of rows in each group,
            unless `sum_over` is specified.
            If `data` is a Series with at most one row for each group, use
            the value of the Series. If `data` is a Series with more than one
            row per group, raise a ValueError.
        'count'
            Count the number of rows in each group.
        'sum'
            Sum the value of the `data` Series, or the DataFrame field
            specified by `sum_over`.
    sum_over : str or None
        If `subset_size='sum'` or `'auto'`, then the intersection size is the
        sum of the specified field in the `data` DataFrame. If a Series, only
        None is supported and its value is summed.
    include_empty_subsets : bool (default=False)
        If True, all possible category combinations will be returned in
        subset_sizes, even when some are not present in data.

    Returns
    -------
    QueryResult
        Including filtered ``data``, filtered and sorted ``subset_sizes`` and
        overall ``category_totals`` and ``total``.

    Examples
    --------
    >>> from upsetplot import query, generate_samples
    >>> data = generate_samples(n_samples=20)
    >>> result = query(data, present="cat1", max_subset_size=4)
    >>> result.category_totals
    cat1    14
    cat2     4
    cat0     0
    dtype: int64
    >>> result.subset_sizes
    cat1  cat2  cat0
    True  True  False    3
    Name: size, dtype: int64
    >>> result.data
                     index     value
    cat1 cat2 cat0
    True True False      0  2.04...
              False      2  2.05...
              False     10  2.55...
    >>>
    >>> # Sorting:
    >>> query(data, min_degree=1, sort_by="degree").subset_sizes
    cat1   cat2   cat0
    True   False  False    11
    False  True   False     1
    True   True   False     3
    Name: size, dtype: int64
    >>> query(data, min_degree=1, sort_by="cardinality").subset_sizes
    cat1   cat2   cat0
    True   False  False    11
           True   False     3
    False  True   False     1
    Name: size, dtype: int64
    >>>
    >>> # Getting each subset's data
    >>> result = query(data)
    >>> result.subsets[frozenset({"cat1", "cat2"})]
                index     value
    cat1  cat2 cat0
    False True False      3  1.333795
    >>> result.subsets[frozenset({"cat1"})]
                        index     value
    cat1  cat2  cat0
    False False False      5  0.918174
                False      8  1.948521
                False      9  1.086599
                False     13  1.105696
                False     19  1.339895
    """

    data, agg = _aggregate_data(data, subset_size, sum_over)
    data = _check_index(data)
    grand_total = agg.sum()
    category_totals = [
        agg[agg.index.get_level_values(name).values.astype(bool)].sum()
        for name in agg.index.names
    ]
    category_totals = pd.Series(category_totals, index=agg.index.names)

    if include_empty_subsets:
        nlevels = len(agg.index.levels)
        if nlevels > 10:
            raise ValueError(
                "include_empty_subsets is supported for at most 10 categories"
            )
        new_agg = pd.Series(
            0,
            index=pd.MultiIndex.from_product(
                [[False, True]] * nlevels, names=agg.index.names
            ),
            dtype=agg.dtype,
            name=agg.name,
        )
        new_agg.update(agg)
        agg = new_agg

    data, agg = _filter_subsets(
        data,
        agg,
        min_subset_size=min_subset_size,
        max_subset_size=max_subset_size,
        max_subset_rank=max_subset_rank,
        min_degree=min_degree,
        max_degree=max_degree,
        present=present,
        absent=absent,
    )

    # sort:
    if sort_categories_by in ("cardinality", "-cardinality"):
        category_totals.sort_values(
            ascending=sort_categories_by[:1] == "-", inplace=True
        )
    elif sort_categories_by == "-input":
        category_totals = category_totals[::-1]
    elif sort_categories_by in (None, "input"):
        pass
    else:
        raise ValueError("Unknown sort_categories_by: %r" % sort_categories_by)
    data = data.reorder_levels(category_totals.index.values)
    agg = agg.reorder_levels(category_totals.index.values)

    if sort_by in ("cardinality", "-cardinality"):
        agg = agg.sort_values(ascending=sort_by[:1] == "-")
    elif sort_by in ("degree", "-degree"):
        index_tuples = sorted(
            agg.index,
            key=lambda x: (sum(x),) + tuple(reversed(x)),
            reverse=sort_by[:1] == "-",
        )
        agg = agg.reindex(
            pd.MultiIndex.from_tuples(index_tuples, names=agg.index.names)
        )
    elif sort_by == "-input":
        agg = agg[::-1]
    elif sort_by in (None, "input"):
        pass
    else:
        raise ValueError("Unknown sort_by: %r" % sort_by)

    return QueryResult(
        data=data, subset_sizes=agg, category_totals=category_totals, total=grand_total
    )
